File: packages/core/src/engine/devOutput.ts
const ANSI_RE = /\x1b\[[0-?]*[ -/]*[@-~]/g;

export type DevOutputOptions = {
  maxTotalLines?: number;
  headLines?: number;
  tailLines?: number;
  maxLineLength?: number;
  testAware?: boolean;
};

const DEFAULTS: Required<DevOutputOptions> = {
  maxTotalLines: 4000,
  headLines: 400,
  tailLines: 400,
  maxLineLength: 480,
  testAware: true,
};

function stripAnsi(text: string): string {
  return text.replace(ANSI_RE, '').replace(/\r\n/g, '\n').replace(/\r/g, '\n');
}

function clampLine(line: string, maxLen: number): string {
  if (line.length <= maxLen) return line;
  const keep = maxLen - 20;
  return `${line.slice(0, keep)} … [+${line.length - keep} chars]`;
}

function collapseDuplicateRuns(lines: string[]): string[] {
  if (lines.length === 0) return [];
  const out: string[] = [];
  let prev = lines[0];
  let count = 1;
  for (let i = 1; i <= lines.length; i++) {
    const cur = lines[i];
    if (cur === prev && i < lines.length) {
      count++;
      continue;
    }
    out.push(count > 1 ? `${prev} (×${count})` : prev);
    if (i < lines.length) {
      prev = cur!;
      count = 1;
    }
  }
  return out;
}

function looksLikeTestOutput(text: string): boolean {
  const head = text.slice(0, 12000);
  return (
    /running\s+\d+\s+tests?\b/i.test(head) ||
    /\btest\s+result:\s/i.test(head) ||
    /^\s*FAIL\s+/m.test(head) ||
    /\bfailures?:\s*$/m.test(head) ||
    /\b\d+\s+passed\b.*\b\d+\s+failed\b/i.test(head) ||
    (/^\s*●\s/m.test(head) && /\btests?\b/i.test(head)) ||
    (/test\s+.*\s+\.\.\.\s+(ok|FAILED)\b/i.test(head) && /\b\d+\s*;\s*\d+\s+failed/i.test(head))
  );
}

function shrinkTestLines(lines: string[]): string[] {
  const n = lines.length;
  const important = (i: number) => {
    const line = lines[i];
    return (
      /FAIL|failed|error\[E|ERROR|Error:|AssertionError|panicked at|test\s+result|failures?:|^\s*●\s|^\s*✕\s|expected|Received|Diff|assert_eq!|panic|thread\s+'/i.test(
        line
      ) ||
      /^\s+at\s/.test(line) ||
      /^running\s+\d+/i.test(line) ||
      /^\s*#\d+\s+/.test(line) ||
      /-->\s/.test(line)
    );
  };
  const kept = new Set<number>();
  const head = Math.min(24, n);
  const tail = Math.min(20, n);
  for (let i = 0; i < head; i++) kept.add(i);
  for (let i = n - tail; i < n; i++) if (i >= 0) kept.add(i);
  for (let i = 0; i < n; i++) if (important(i)) kept.add(i);
  const sorted = [...kept].sort((a, b) => a - b);
  let out = sorted.map((i) => lines[i]);
  const max = 450;
  if (out.length > max) {
    const drop = out.length - max + 1;
    out = [...out.slice(0, max - 1), `... [+${drop} linhas omitidas]`];
  }
  return out;
}

function truncateMiddle(lines: string[], head: number, tail: number): string[] {
  if (lines.length <= head + tail + 8) return lines;
  const omitted = lines.length - head - tail;
  return [...lines.slice(0, head), `… [${omitted} linhas omitidas] …`, ...lines.slice(-tail)];
}

export function devOutputPipeline(text: string, options: DevOutputOptions = {}): { output: string; noiseRemoved: number } {
  const o = { ...DEFAULTS, ...options };
  const raw = text.length;
  let s = stripAnsi(text);
  if (!s.trim()) return { output: '', noiseRemoved: 0 };
  let lines = s.split('\n');
  if (o.testAware && looksLikeTestOutput(s)) {
    lines = shrinkTestLines(lines);
  }
  lines = lines.map((l) => clampLine(l, o.maxLineLength));
  lines = collapseDuplicateRuns(lines);
  if (lines.length > o.maxTotalLines) {
    lines = truncateMiddle(lines, o.headLines, o.tailLines);
  }
  const out = lines.join('\n').trimEnd();
  const noiseRemoved = raw > 0 ? Math.max(0, Math.floor(((raw - out.length) / raw) * 100)) : 0;
  return { output: out, noiseRemoved };
}

File: crates/pith-core/src/dev_output.rs
use crate::types::DevOutputResult;
use regex::Regex;

#[derive(Debug, Clone)]
pub struct DevOutputOptions {
    pub max_total_lines: usize,
    pub head_lines: usize,
    pub tail_lines: usize,
    pub max_line_length: usize,
    pub test_aware: bool,
}

impl Default for DevOutputOptions {
    fn default() -> Self {
        Self {
            max_total_lines: 4000,
            head_lines: 400,
            tail_lines: 400,
            max_line_length: 480,
            test_aware: true,
        }
    }
}

fn strip_ansi(text: &str) -> String {
    let re = Regex::new(r"\x1b\[[0-?]*[ -/]*[@-~]").expect("valid regex");
    re.replace_all(text, "").replace("\r\n", "\n").replace('\r', "\n")
}

fn clamp_line(line: &str, max_len: usize) -> String {
    if line.len() <= max_len {
        return line.to_string();
    }
    let keep = max_len.saturating_sub(20);
    format!("{} … [+{} chars]", &line[..keep], line.len().saturating_sub(keep))
}

fn collapse_duplicate_runs(lines: Vec<String>) -> Vec<String> {
    if lines.is_empty() {
        return lines;
    }
    let mut out = Vec::new();
    let mut prev = &lines[0];
    let mut count = 1usize;
    for line in lines.iter().skip(1) {
        if line == prev {
            count += 1;
            continue;
        }
        out.push(if count > 1 { format!("{prev} (×{count})") } else { prev.clone() });
        prev = line;
        count = 1;
    }
    out.push(if count > 1 { format!("{prev} (×{count})") } else { prev.clone() });
    out
}

fn looks_like_test_output(text: &str) -> bool {
    let head = &text[..text.len().min(12000)];
    Regex::new(r"(?i)running\s+\d+\s+tests?|test\s+result:|^\s*FAIL\s+|failures?:\s*$")
        .expect("valid regex")
        .is_match(head)
}

fn shrink_test_lines(lines: Vec<String>) -> Vec<String> {
    let important = Regex::new(r"(?i)FAIL|failed|error\[E|ERROR|Error:|AssertionError|panicked at|test\s+result|failures?:|expected|Received|Diff|panic")
        .expect("valid regex");
    let mut kept = Vec::new();
    for (idx, line) in lines.iter().enumerate() {
        if idx < 24 || idx + 20 >= lines.len() || important.is_match(line) || line.trim_start().starts_with("at ") {
            kept.push(line.clone());
        }
    }
    if kept.len() > 450 {
        let drop = kept.len() - 449;
        let mut trimmed = kept[..449].to_vec();
        trimmed.push(format!("... [+{drop} linhas omitidas]"));
        return trimmed;
    }
    kept
}

fn truncate_middle(lines: Vec<String>, head: usize, tail: usize) -> Vec<String> {
    if lines.len() <= head + tail + 8 {
        return lines;
    }
    let omitted = lines.len() - head - tail;
    [
        lines[..head].to_vec(),
        vec![format!("… [{omitted} linhas omitidas] …")],
        lines[lines.len() - tail..].to_vec(),
    ]
    .concat()
}

pub fn dev_output_pipeline(text: &str, options: Option<DevOutputOptions>) -> DevOutputResult {
    let o = options.unwrap_or_default();
    let raw = text.len();
    let mut s = strip_ansi(text);
    if s.trim().is_empty() {
        return DevOutputResult { output: String::new(), noise_removed: 0 };
    }

    let mut lines = s.lines().map(str::to_string).collect::<Vec<_>>();
    if o.test_aware && looks_like_test_output(&s) {
        lines = shrink_test_lines(lines);
    }
    lines = lines.into_iter().map(|l| clamp_line(&l, o.max_line_length)).collect();
    lines = collapse_duplicate_runs(lines);
    if lines.len() > o.max_total_lines {
        lines = truncate_middle(lines, o.head_lines, o.tail_lines);
    }

    s = lines.join("\n").trim_end().to_string();
    let noise_removed = if raw > 0 {
        ((raw.saturating_sub(s.len())) * 100) / raw
    } else {
        0
    };
    DevOutputResult { output: s, noise_removed }
}
